### PR TITLE
Fix: Submit duplicate sequencing files

### PR DESCRIPTION
**Pull request by leoraba**
# Description
This PR intercepts duplicates in submitting sequencing files metadata and returns the corresponding error

The submit endpoint will return a 200 OK response  with status **INVALID_SUBMISSION**, and a message indicating what files are duplicated.

Example:
```json
{
    "submissionId": 108,
    "status": "INVALID_SUBMISSION",
    "submissionManifest": [],
    "batchErrors": [
        {
            "message": "The following files have already been submitted for this submission: 21-WW-BAR-01-Illumina-20.tar.xz",
            "type": "INCORRECT_SECTION",
            "batchName": "Submission 108"
        }
    ]
}
```

## More details
- **New Drizzle Migration:** Added `md5_sum` column to the `record_analysis_map` table to store the MD5 checksum of submitted sequencing files. (`0003_add_md5_sum_column_map_table.sql`)
- **Unit Tests:** Added coverage for `findDuplicateSequencingMetadata` and `findAlreadySubmittedFiles`. (`sequencingPayload.test.ts`)

## Related Issue
- https://github.com/imicroseq/roadmap/issues/175

### CODE REVIEW

File: src/submission/submissionHandler.ts [hunk 1]
```diff
@@ -262,6 +262,21 @@ export async function handleSubmission({
 	};
 }
 
+/**
+ * Finds duplicate files in the sequencing metadata values based on their md5sum
+ */
+const findDuplicateSequencingMetadata = (
+	sequencingMetadataValues: SequencingMetadataType[],
+): SequencingMetadataType[] => {
```
Comment (justincorrigible) [inline src/submission/submissionHandler.ts]: A first-time upload with two rows sharing the same `fileMd5sum` still creates two Song analyses for one file, the exact bug #175 describes, just via the initial-upload path instead of the append path:
this duplicate check only runs in `handleSequencingMetadataSubmission`.

`handleSubmission` takes the same `sequencingMetadataValues` and feeds it through the same `buildSequencingFilesMetadata` → `submitSongPayload` chain, but never calls `findDuplicateSequencingMetadata` or checks against existing files. 

Since both functions funnel through `buildSequencingFilesMetadata` (fileValidation.ts) and `submitSongPayload` before reaching Song, moving the dedup logic into one of those shared functions would close both paths at once instead of duplicating the check into `handleSubmission` too.

File: src/submission/submissionHandler.ts [hunk 2]
```diff
@@ -300,6 +315,43 @@ export async function handleSequencingMetadataSubmission({
...
+					message: `The following files have duplicate md5sum values: ${duplicateMetadataFiles.map((metadata) => metadata.fileName).join(', ')}`,
+					type: BATCH_ERROR_TYPE.INCORRECT_SECTION,
+					batchName,
+				},
+			],
+		};
+	}
+
+	// Submission valiation - Find any duplicates againt submisison file metadata
+	const existingFiles = await buildSubmissionFileMetadata(organization, submissionId);
```
Comment (justincorrigible) [inline src/submission/submissionHandler.ts]: this check is a "read, then decide" with no lock or transaction around it:
two concurrent requests for the same submission and file both read `existingFiles` before either has written, both pass, and both get submitted to Song... that's the same duplicate this PR is meant to prevent, just under concurrency.
i.e. `record_analysis_map` only has non-unique indexes on `submission_id` and `system_id`, nothing on the file's `md5sum` or identifier...

worth a unique constraint on the DB side as the actual backstop, with this check staying as the fast-path rejection

File: src/submission/submissionHandler.ts [hunk 3]
```diff
@@ -300,6 +315,43 @@ export async function handleSequencingMetadataSubmission({
...
+	const alreadySubmittedFiles = sequencingMetadataValues.filter((metadata) =>
+		existingFiles.find((existing) => existing.md5Sum === metadata.fileMd5sum),
+	);
```
Comment (justincorrigible) [inline src/submission/submissionHandler.ts]: Three gaps here:

1. `existing.md5Sum === metadata.fileMd5sum` is a strict, case-sensitive comparison with no normalization: two submissions of the same file with differently-cased hashes won't be caught.
2. `fileMd5sum` has no non-empty constraint upstream: two genuinely different files both submitted with an empty or unset `md5sum` get grouped as duplicates of each other here and in `findDuplicateSequencingMetadata` above, a false positive that blocks a legitimate submission.
3. This doesn't check `existing.isUploaded`: a file staged but not yet published can get flagged as already submitted when it isn't actually finalized, blocking a legitimate resubmission or correction before publish.

File: src/submission/submissionHandler.ts (hunk 2 above)
Comment (justincorrigible) [reply src/submission/submissionHandler.ts]: on a separate note: this is a serial loop doing one Song `getAnalysisById` call per existing file, so a submission that later fails validation now pays that full cost on every attempt instead of only on the success path like before this PR (old line 357).

`buildSubmissionFileMetadata` now runs unconditionally right after the duplicate-md5 check, before `getSubmissionDetailsById`, `buildSequencingFilesMetadata`, and `submitSongPayload`, any of which can still reject the request.

File: src/submission/submissionHandler.ts (hunk 3 above)
Comment (justincorrigible) [reply src/submission/submissionHandler.ts]: also separately here:
`existingFiles.find(...)` inside `.filter(...)` is O(n·m). 

`fileValidation.ts` already uses a Set for the equivalent membership check a few lines over.
building `const existingMd5Sums = new Set(existingFiles.map(f => f.md5Sum))` once and using `.has()` would match that pattern and drop this to O(n+m).

File: src/submission/submissionHandler.ts [hunk 4]
```diff
@@ -262,6 +262,21 @@ export async function handleSubmission({
...
+ * Finds duplicate files in the sequencing metadata values based on their md5sum
+ */
+const findDuplicateSequencingMetadata = (
+	sequencingMetadataValues: SequencingMetadataType[],
+): SequencingMetadataType[] => {
+	const metadataFileNamesByMd5sum = sequencingMetadataValues.reduce(
+		(fileNamesByMd5sum, metadata) =>
+			fileNamesByMd5sum.set(metadata.fileMd5sum, [...(fileNamesByMd5sum.get(metadata.fileMd5sum) ?? []), metadata]),
+		new Map<string, SequencingMetadataType[]>(),
+	);
```
Comment (justincorrigible) [inline src/submission/submissionHandler.ts]: `fileNamesByMd5sum.set(key, [...(get(key) ?? []), metadata])` copies the whole bucket on every insert, quadratic for a batch with a large duplicate cluster... a count-first pass (`Map<string, number>`), then filtering the original array by count > 1, avoids the copies and states the actual "does this `md5sum` appear more than once" more directly.

File: src/submission/submissionHandler.ts [hunk 5]
```diff
@@ -300,6 +315,43 @@ export async function handleSequencingMetadataSubmission({
...
+	}
+
+	// Submission valiation - Find any duplicates againt submisison file metadata
+	const existingFiles = await buildSubmissionFileMetadata(organization, submissionId);
+	const alreadySubmittedFiles = sequencingMetadataValues.filter((metadata) =>
+		existingFiles.find((existing) => existing.md5Sum === metadata.fileMd5sum),
+	);
+
+	if (alreadySubmittedFiles.length) {
+		return {
```
Comment (justincorrigible) [inline src/submission/submissionHandler.ts]: Worth a console log line for consistency, and so a rejected submission shows up in the logs the same way the existing ones do... other rejection paths in this function log before returning. 
These two new ones don't (here and the one in line 322.)

Comment (justincorrigible) [reply src/submission/submissionHandler.ts]: semi-related note: all of these result objects could be abstracted into a `buildErrorResult(submissionId, batchName, message)` to avoid repetition.

File: src/submission/submissionHandler.ts [hunk 6]
```diff
@@ -300,6 +315,43 @@ export async function handleSequencingMetadataSubmission({
 		};
 	}
 
+	// Input validation - Find duplicate md5sums in senquencing metadata input
```
Comment (justincorrigible) [inline src/submission/submissionHandler.ts]:
```suggestion
	// Input validation - Find duplicate md5sums in sequencing metadata input
```

File: src/submission/submissionHandler.ts [hunk 7]
```diff
@@ -300,6 +315,43 @@ export async function handleSequencingMetadataSubmission({
...
+				{
+					message: `The following files have duplicate md5sum values: ${duplicateMetadataFiles.map((metadata) => metadata.fileName).join(', ')}`,
+					type: BATCH_ERROR_TYPE.INCORRECT_SECTION,
+					batchName,
+				},
+			],
+		};
+	}
+
+	// Submission valiation - Find any duplicates againt submisison file metadata
```
Comment (justincorrigible) [inline src/submission/submissionHandler.ts]:
```suggestion
	// Submission validation - Find any duplicates against submission file metadata
```

**Review by justincorrigible — CHANGES_REQUESTED** [review]
left a few comments with some critical items. beyond that, a few `node:test` cases (duplicate md5, no duplicates, empty and case-variant md5 per the notes above) would double as executable documentation of the edge cases before this ships.

File: src/submission/submissionHandler.ts (hunk 1 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: good catch, I'll move the calling `findDuplicateSequencingMetadata` validation at controller level, this makes sure to validate the request before any submission logic

File: src/submission/submissionHandler.ts (hunk 3 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: applied following changes as suggested in **Gap # 1**:
1. Isolated this logic in it's own function `findAlreadySubmittedFiles ` 
2. Normalize both md5sum strings using `.toLowerCase()` 
3. Build a `Set` to store unique md5 strings
4. Using `.has()` to match md5 strings.

File: src/db/schemas/record_analysis_map.ts [hunk 8]
```diff
@@ -28,6 +28,7 @@ export const submissionFiles = schema.table(
 		submission_id: integer().notNull(),
 		record_identifier: varchar({ length: 255 }).notNull(),
 		analysis_id: varchar({ length: 255 }).notNull(),
+		md5_sum: varchar({ length: 255 }),
```
Comment (leoraba) [inline src/db/schemas/record_analysis_map.ts:32]: new column added to track the md5 checksum!

File: src/submission/submissionHandler.ts (hunk 5 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: created the function `createSubmissionError` to be used consistently to log and return error response

File: src/submission/submissionHandler.ts (hunk 2 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: Replaced calling `buildSubmissionFileMetadata` by a new function `fetchSubmissionFilesBySubmissionId` which only calls the database to retrieve submission files metadata avoiding to call Song.

It was added a new column `md5_sum` in table `record_analysis_map` to store the MD5 checksum in submission service database.

File: src/submission/submissionHandler.ts (hunk 3 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: Responding:
**Gap # 2**: Empty MD5 sums are ignored, so no duplicate error is returned for them.

**Gap # 3:** The `isUpload` field is currently ignored. A duplicate error is returned whenever the MD5 sum matches, regardless of the `isUpload` value.

File: src/submission/submissionHandler.ts (hunk 4 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: this function was refactored as suggested, using a `Map<string, number>` object to keep the count based on md5sum. the function was later renamed to `findDuplicateSequencingMetadata` and  relocated to `sequencingPayload.ts`

File: src/submission/submissionHandler.ts (hunk 6 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: updated as suggested 👍

File: src/submission/submissionHandler.ts (hunk 7 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: type fixed

File: src/submission/sequencingPayload.test.ts [hunk 9]
```diff
@@ -20,7 +20,103 @@
...
+) => ({
+	id: 1,
+	submission_id: submissionId,
+	record_identifier: recordIdentifier,
+	analysis_id: analysisId,
+	md5_sum: md5Sum,
+	created_at: Date.now().toString(),
+});
+
+suite('findDuplicateSequencingMetadata', () => {
```
Comment (leoraba) [inline src/submission/sequencingPayload.test.ts]: unit tests to cover function `findDuplicateSequencingMetadata` and `findAlreadySubmittedFiles`

**Comment by leoraba** [comment]
PR updated based on the received feedback and ready for code review.

File: src/submission/submissionHandler.ts (hunk 1 above)
Comment (justincorrigible) [reply src/submission/submissionHandler.ts]: moving it to the controller is a better fix than what I suggested: it covers both paths from one place and runs before any submission logic, so neither handler can be reached with an intra-batch duplicate.
Confirmed in `submit.ts:63-77`

the other half of this thread is still open though. `findAlreadySubmittedFiles` only runs in `handleSequencingMetadataSubmission`, so the cross-submission check still doesn't cover the initial-upload path... Concretely: submit file + metadata for file A (row lands in `record_analysis_map` with A's md5), then submit a second file + metadata for the same A. That second request goes through `handleSubmission`, which never calls `findAlreadySubmittedFiles`, so it registers a second Song analysis for A.


---
whether that's in scope here is your call, but it's the same shape as the intra-batch gap this thread started with, so worth an explicit decision rather than leaving it implicit.

File: src/submission/submissionHandler.ts (hunk 2 above)
Comment (justincorrigible) [reply src/submission/submissionHandler.ts]: the `fetchSubmissionFilesBySubmissionId` swap is a genuinely better answer than reordering the call, which is all I'd suggested: one DB query instead of N serial Song round trips, and it now runs before `getSubmissionDetailsById` so the cheap check comes first. 👍 
That also fully closes the efficiency comment.

however... the race itself is still open. the new `md5_sum` column makes it cheap to close now, which it wasn't before: `record_analysis_map` still has no unique constraint (0003 adds a plain nullable column, and the snapshot shows `uniqueConstraints: {}`), so two concurrent requests for the same file both read, both pass the check, and both submit to Song.

one wrinkle if you do add it: because `md5_sum` is written as `|| ''` (see the separate comment on that line), a plain unique index on `(submission_id, md5_sum)` would reject a second file with no md5 at all. A partial index avoids that:

```sql
CREATE UNIQUE INDEX ... ON submission.record_analysis_map (submission_id, md5_sum)
  WHERE md5_sum IS NOT NULL AND md5_sum <> '';
```

fine by me if this is something you wanna hit in a follow-up PR, but I'd just rather it be deliberate. might as well fix it here. your call, of course

File: src/submission/submissionHandler.ts (hunk 3 above)
Comment (justincorrigible) [reply src/submission/submissionHandler.ts]: remaining attention needed

- Gap 2 (empty md5): the fix is correct, ignoring empties is right for this function since it's the only way to avoid the false positive. It does surface a question one layer up though: `fileMd5sum` is `z.string()` (why you no `zod.string`!? lol) in `submitRequest.ts` with no minimum length, so a submission where every md5 is empty now passes both duplicate checks with zero protection and no error. "Ignored" is the correct behaviour for the matcher, but it might be worth rejecting an empty `fileMd5sum` at the schema level, so the state is validated rather than silently skipped.

- Gap 3 (`isUploaded`): worth noting the answer changed shape rather than being a choice now. With the switch to `fetchSubmissionFilesBySubmissionId`, `isUploaded` isn't fetched at all any more, it came from `buildSubmissionFileMetadata`'s Song `analysisState === 'PUBLISHED'` check. So it isn't being ignored, it's structurally unavailable at that point.

... that's "mostly" fine, and arguably better even: a row only exists in `record_analysis_map` after Song accepted the submission, so a match really does mean a duplicate. The narrow case left is a user who needs to resubmit the same file's metadata against a submission that isn't committed or published yet, for instance to correct something: if that's not a workflow you want to support, no change needed, just confirming it's intended rather than just "incidental"

File: src/submission/submissionHandler.ts (hunk 5 above)
Comment (justincorrigible) [reply src/submission/submissionHandler.ts]: `createSubmissionError` is the right shape and covers the logging point and the `buildErrorResult` suggestion from [3798695732](https://github.com/imicroseq/submission-service/pull/27#discussion_r3798695732) in one go. 👍 

small note: the intra-batch duplicate rejection moved to the controller in the process, and that site doesn't log. `respondWithInvalidSubmission` has no logging of its own, so a rejected duplicate on the initial-upload path now leaves no trace, which is the same gap this thread opened with, just at its new location. A `logger.info` alongside the return would keep it consistent

File: src/submission/sequencingPayload.test.ts (hunk 9 above)
Comment (justincorrigible) [reply src/submission/sequencingPayload.test.ts]: good coverage, and it's targeted at exactly the edge cases that were raised rather than just the happy path. the case-insensitivity and empty-md5 tests are the two I'd have asked for.

three notes:

1. The fixtures aren't type-checked by anything. `tsconfig.json` excludes `**/*.test.ts`, and `tsx` strips types without checking them, so nothing verifies the fixture matches the real row shape. It currently doesn't: `mappingSequencingMetadata` omits `system_id` and sets `created_at` to a string, where `SelectSubmissionFile` has `system_id: string | null` and `created_at: Date`. Harmless at runtime, but it means the test's realism isn't guaranteed by anything, and a future change to the row shape won't fail here.

2. `exsistingFiles` is misspelled in both suites.

3. The one link in the chain still untested is the write: `md5_sum: record.files[0]?.fileMd5sum || ''` in `submitSongPayload`. Both new functions are covered, but if that write is ever wrong the checks silently match nothing, and nothing would catch it.

File: src/db/schemas/record_analysis_map.ts (hunk 8 above)
Comment (justincorrigible) [reply src/db/schemas/record_analysis_map.ts:32]: right call, and it's what makes the check cheap. Two things about it worth deciding on:

1. Existing rows get `NULL`, and 0003 doesn't backfill... since `findAlreadySubmittedFiles` skips falsy md5s, every file submitted before this deploys is invisible to the new check, so on day one a user can resubmit any pre-existing file and it won't be caught. The value is recoverable from Song (`getAnalysisById(...).files[0].fileMd5sum`, which is what `buildSubmissionFileMetadata` already reads), so a one-off backfill looks feasible. Worth knowing whether that's planned or whether the check is knowingly forward-only.

2. The column is nullable while the write path uses `|| ''`, so "no md5" ends up represented two ways: `NULL` for pre-existing rows, `''` for new ones. Nothing breaks today since the falsy check catches both, but it's an ambiguity that gets harder to remove later and it blocks a plain unique index (see the race thread).

File: src/controllers/submission/submit.ts [hunk 10]
```diff
@@ -60,6 +61,21 @@ export const submit = validateRequest(
 				});
 			}
 
+			// Input validation - Find duplicate md5sums in sequencing metadata input
+			if (sequencingMetadataValues && sequencingMetadataValues.length > 0) {
+				const duplicateMetadataFiles = findDuplicateSequencingMetadata(sequencingMetadataValues);
+
+				if (duplicateMetadataFiles.length) {
+					return respondWithInvalidSubmission(res, undefined, [
```
Comment (justincorrigible) [inline src/controllers/submission/submit.ts]: two small consistency points now that this check lives here rather than in the handler:

1. `batchName` is `submissionFile?.originalname || ''`, so on the append path (no file) the client gets an empty `batchName`, where the same class of error from `handleSequencingMetadataSubmission` reports `Submission {id}`. Worth making these match, since a client parsing `batchName` sees two different things for two duplicate errors.

2. `submissionId` is passed as `undefined`. On the append path an active submission does exist, it just hasn't been fetched yet at this point in the flow. The PR description's example response shows a populated `submissionId`, so the two duplicate-error classes now return different response shapes: **Not wrong**, but worth being deliberate about.

**Review by justincorrigible — COMMENTED** [review]
Great work! we're so much closer with these changes 🎉 
and the `fetchSubmissionFilesBySubmissionId` swap and the controller-level relocation are both better than what I'd suggested, and the new tests cover the edge cases that prompted the review rather than just the happy path.

Two things I'd want settled before merge, both **decisions rather than code**:

1. Are pre-existing `record_analysis_map` rows getting a `md5_sum` backfill? Without one the check is forward-only, so it won't catch duplicates of anything already submitted.
2. Is the initial-upload path (`handleSubmission`) meant to stay outside the cross-submission "already submitted" check? Intra-batch is covered on both paths now, cross-submission only on the append path.

---
Everything else is small: the empty-string versus NULL representation for `md5_sum`, a missing log line on the relocated check, `batchName`/`submissionId` shape differences between the two duplicate errors, and the test fixtures not being type-checked.

The concurrency gap (no unique constraint) is still there and is now cheap to close given the new column, but is reasonable to defer to its own PR if that's the intent.

File: src/submission/submissionHandler.ts (hunk 2 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: the uniqueness for md5 sum is optional and manageable by setting the environment variable `SEQUENCING_SUBMISSION_ALLOW_DUPLICATES`.

For that reason a `UNIQUE INDEX` as suggested wouldn't work, instead a **standard (non-unique) index** was added into the `00003` migration, so a query to look up for existing md5_sums can be used when applicable.

File: src/submission/submissionHandler.ts (hunk 3 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: **Gap 1: **
 -  Refactored this function and moved the logic to functions `findPreviouslySubmittedRecordIdentifierMatches` and `findPreviouslySubmittedMd5SumMatches`
**Gap 2: **
  - alias zod import as 'zod' to avoid naming conflicts in file `submitRequest.ts`
  - updated the schema request to accept this format only `zod.string().regex(/^[a-fA-F0-9]{32}$/` The `fileMd5sum` is required.
 **Gap 3: **
  - `isUploaded` unavailable after refactoring.

File: src/submission/submissionHandler.ts (hunk 5 above)
Comment (leoraba) [reply src/submission/submissionHandler.ts]: added missing log in to the function `respondWithInvalidSubmission `
Example:
`Invalid submission 101: ["The following files have duplicate MD5 sums: 21-WW-BAR-01-Illumina-24.tar.xz"]`

File: src/db/schemas/record_analysis_map.ts (hunk 8 above)
Comment (leoraba) [reply src/db/schemas/record_analysis_map.ts:32]: 1. The `0003` drizzle migration won't do any backfill, although possible doable in another PR if required, focusing now only in forward only submission.
2. For new submission it is restricted to pass fileMd5sum empty, added a required MD5 format validation at the request input validation to make sure every file now contains a valid value. Although for previous submision files md5 is null and ignored to find duplicity.

File: src/submission/sequencingPayload.test.ts (hunk 9 above)
Comment (leoraba) [reply src/submission/sequencingPayload.test.ts]: fixed the 3 notes:
1. updated `tsconfig.json` to include TS checks in the tests files, and solved the type errors mentioned.
2. fixed typo
3. Integrations tests are not included in this PR as are away of the scope of the main issue to solve, furthermore `md5sum` is required and being validation at request validation, so it should always contain a valid MD5 value.

File: src/controllers/submission/submit.ts (hunk 10 above)
Comment (leoraba) [reply src/controllers/submission/submit.ts]: 1. Removed the wrong assignment of `batchName` to `Submission {id}`,  the `batchName` is meant to indicate the `.csv` attached in the payload, which in some cases will be empty as it's possible to have a  Submission requests without a `.csv` and to only include `.tar.xz` files. Refactored the code to be consistent as described.
2.  Indeed the `submissionId` could be `undefined` when there is no actual submission but an error in the payload is found, no current change was made about this but raised awareness.

**Comment by leoraba** [comment]
Updates this PR to solve feedback, refactored helper functions to validate duplicity, now it could detect duplicate files based on MD5sums or sampleIds.
**PR is ready for code review.**

File: tsconfig.json [hunk 11]
```diff
@@ -25,5 +25,5 @@
 		"skipLibCheck": true /* Skip type checking all .d.ts files. */
 	},
 	"include": ["./src/**/*.ts", "register.ts", "./src/templates/*.json"],
-	"exclude": ["dist", "node_modules", "**/*.spec.ts", "**/*.test.ts", "*.config.js"]
+	"exclude": ["dist", "node_modules", "*.config.js"]
```
Comment (justincorrigible) [inline tsconfig.json:28]: Regression: removing the test patterns here may fix a type-checking gap, but that exclusion was doing two jobs and this drops both: test files now compile into the build output.

With `outDir: "dist"`, `noEmit: false`, `include: ["./src/**/*.ts", ...]` and `build:compile` being `rimraf dist && tsc`, every `src/**/*.test.ts` is now emitted to `dist/` and ships in the image (big no no). 
Keeping tests out of the build was the exclusion's original purpose; keeping them out of type-checking was an accidental side effect of sharing one mechanism, and only that side effect needed removing.

The split gets both: keep `**/*.test.ts` and `**/*.spec.ts` in `exclude`, add a `tsconfig.test.json` extending this one with those files included and `noEmit: true`, and add a `typecheck` script running `tsc --noEmit && tsc --noEmit -p tsconfig.test.json`.

Confirms either way in one line: `pnpm build:compile && find dist -name '*.test.js'`.

File: src/db/schemas/record_analysis_map.ts [hunk 12]
```diff
@@ -28,10 +29,19 @@ export const submissionFiles = schema.table(
...
 	},
-	(table) => [index('submission_id_idx').on(table.submission_id), index('system_id_idx').on(table.system_id)],
+	(table) => [
+		index('submission_id_idx').on(table.submission_id),
+		index('system_id_idx').on(table.system_id),
+		index('md5_sum_idx')
+			.on(table.md5_sum, table.submission_id, table.record_identifier)
+			.where(
+				sql`${table.md5_sum} IS NOT NULL AND ${table.md5_sum} <> '' AND ${table.system_id} IS NOT NULL AND ${table.system_id} <> ''`,
+			),
```
Comment (justincorrigible) [inline src/db/schemas/record_analysis_map.ts]: this index probably won't be used by the query it was added for, because its predicate asks for more than the query can promise.

Postgres only chooses a partial index when it can prove the query's `WHERE` implies the index's. The query is `inArray(md5_sum, md5sums)` plus, only when `commitedOnly` (typo lol) is true, `isNotNull(system_id)` (`fileRepository.ts:43-46`). Against the four conditions here:

- `md5_sum IS NOT NULL` is implied by the `IN`. Fine.
- `system_id IS NOT NULL` matches directly, but only on the `commitedOnly` path.
- `md5_sum <> ''` isn't provable: the values arrive as a parameterized array, so the planner can't know none of them is the empty string.
- `system_id <> ''` isn't provable at all, since the query never mentions it.

Two of four unprovable means the index gets skipped, and on the `commitedOnly: false` path there's no `system_id` predicate either.

Dropping both `<> ''` conditions should make it usable: `IS NOT NULL` is already implied by the `IN`, and non-emptiness is now guaranteed for new rows by the md5 regex in `submitRequest.ts`. If empty-string hygiene is still worth enforcing, a `CHECK` constraint states it without making the index unreachable.

The column list is a separate mismatch: the index is on `(md5_sum, submission_id, record_identifier)` while the query filters on `md5_sum` and `system_id`. `system_id` isn't in the index and neither trailing column is filtered on, so `(md5_sum, system_id)` would match the actual access pattern.

Worth an `EXPLAIN` on a table large enough that a seq scan isn't preferred anyway, rather than taking my reasoning for it.

File: src/submission/sequencingPayload.ts [hunk 13]
```diff
@@ -26,6 +30,151 @@ import { convertRecordToPayload, prefixKeys } from './populateTemplate.js';
...
+		const identifier = getIdentifierFromFileName(metadata.fileName).toLowerCase();
+		return existingIdentifiers.has(identifier);
+	});
+};
+
+export const findPreviouslySubmittedMd5SumMatches = (
+	existingFiles: SelectSubmissionFile[],
+	sequencingMetadataValues: SequencingMetadataType[],
+): SequencingMetadataType[] => {
+	const existingMd5sums = new Set(existingFiles.flatMap(({ md5_sum }) => (md5_sum ? [md5_sum.toLowerCase()] : [])));
```
Comment (justincorrigible) [inline src/submission/sequencingPayload.ts:97]: On integration tests being out of scope: agreed, no argument there. The supporting reason needs one correction though, because it changes what's untested.
However, "`md5sum` is required and validated, so it should always contain a valid value" only holds for new submissions. Every row predating this migration has `md5_sum` as `NULL`, which is exactly why the falsy-skip branches in `findPreviouslySubmittedMd5SumMatches` and `findDuplicateMd5SumsInMetadata` exist. So those branches are live code exercised by real production data, not dead code kept for safety.

A single case with a `NULL`-md5 existing row would cover it, and it's the branch most likely to matter on day one given the forward-only backfill decision.

File: src/repository/fileRepository.ts [hunk 14]
```diff
@@ -26,6 +26,33 @@ import { type InsertSubmissionFile, type SelectSubmissionFile, submissionFiles }
...
 	return {
+		/**
+		 * Retrieves submission files by their MD5 checksums
+		 * @param md5sums Array of MD5 checksums to search for
+		 * @param commitedOnly If true, only returns files that have been committed
+		 * @returns Array of matching submission files
+		 */
+		getSubmissionFilesByMd5sum: async (
+			md5sums: string[],
+			commitedOnly: boolean = false,
```
Comment (justincorrigible) [inline src/repository/fileRepository.ts]:
```suggestion
			committedOnly: boolean = false,
```

File: src/service/fileService.ts [hunk 15]
```diff
@@ -47,6 +47,18 @@ export const fetchSubmissionFilesBySubmissionId = async (submissionId: number) =
 	return submissionFiles;
 };
 
+/**
+ * Retrieves files by their MD5 checksums
+ * @param md5sums An array of MD5 checksums to search for
+ * @param commitedOnly If true, only returns files that have been committed
+ * @returns A list of submission files matching the given MD5 checksums
+ */
+export const fetchSubmissionFilesByMd5sum = async (md5sums: string[], commitedOnly: boolean = false) => {
```
Comment (justincorrigible) [inline src/service/fileService.ts]:
```suggestion
export const fetchSubmissionFilesByMd5sum = async (md5sums: string[], committedOnly: boolean = false) => {
```

File: tsconfig.json (hunk 11 above)
Comment (leoraba) [reply tsconfig.json:28]: added a new file `tsconfig.build.json` which this file and excludes the tests files.

File: tsconfig.build.json [hunk 16]
```diff
@@ -0,0 +1,4 @@
+{
+	"extends": "./tsconfig.json",
+	"exclude": ["dist", "node_modules", "**/*.spec.ts", "**/*.test.ts", "*.config.js"]
```
Comment (leoraba) [inline tsconfig.build.json:3]: exclude tests files during build

File: src/db/schemas/record_analysis_map.ts (hunk 12 above)
Comment (leoraba) [reply src/db/schemas/record_analysis_map.ts]: updated this index to columns `md5_sum` and `system_id`  where `md5_sum IS NOT NULL`, removing the  `md5_sum <> ''`, `system_id <> ''`  and `system_id IS NOT NULL` conditions, this query is now more useful finding md5 duplicates.

File: src/service/fileService.ts (hunk 15 above)
Comment (leoraba) [reply src/service/fileService.ts]: typo fixed

File: src/repository/fileRepository.ts (hunk 14 above)
Comment (leoraba) [reply src/repository/fileRepository.ts]: typo fixed

File: src/submission/sequencingPayload.ts (hunk 13 above)
Comment (leoraba) [reply src/submission/sequencingPayload.ts:97]: added missing unit test to cover `NULL` md5sums scenario

**Comment by leoraba** [comment]
PR updated based on feedback received, adding 2 new tech debts for future development. 
Ready for code review.

File: tsconfig.json (hunk 11 above)
Comment (justincorrigible) [reply tsconfig.json:28]: splitting into `tsconfig.build.json` is a cleaner arrangement than what I suggested, and it fixes the emission problem properly: base config for development, narrow config for the build.

there's still one loose end though... the base `tsconfig.json` is now the only config that includes the test files, and nothing in `scripts` runs it: `build:compile` points at `tsconfig.build.json`, and `test` goes through `tsx`, which strips types without checking them. so the test files are type-checked by the editor and by nothing that can be run on demand or in CI, which is close to where we started, just for a better reason.

a one-liner solves this: `"typecheck": "tsc --noEmit"`. That uses the base config, so it covers `src` and the tests together, and it gives the CI workflow already tracked in `.dev/tech-debt.md` something concrete to call alongside `pnpm test` and `pnpm lint`.

Comment (leoraba) [reply tsconfig.json:28]: added `"typecheck"` script  as suggested, it can be used on demand and by default uses `tsconfig.json`

**Review by justincorrigible — APPROVED** [review]

**Comment by justincorrigible** [comment]
ready to squash away 👍
